content/en/blog/_posts/2024-09-20-kubeadm-v1beta4.md

```diff
@@ -0,0 +1,99 @@
+and how to migrate from v1beta3 to v1beta4.
+
+Details about v1beta4 can be found in the
+[API doc]((/docs/reference/config-api/kubeadm-config.v1beta4/)).
+The documentation is maintained in the
+[code](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/v1beta4/doc.go)
+and generated for the kubernetes.io website.
+
+### A list of changes since v1beta3
+
+This version improves on the [v1beta3](/docs/reference/config-api/kubeadm-config.v1beta3/)
+format by fixing some minor issues and adding a few new fields.
+
+To put it simply,
+
+- Two new API types: ResetConfiguration and UpgradeConfiguration
+- For InitConfiguration and JoinConfiguration, `dryRun` mode and `nodeRegistration.imagePullSerial` are supported
+- For ClusterConfiguration, we add field including `certificateValidityPeriod`,
+`caCertificateValidityPeriod`, `encryptionAlgorithm`, `dns.disabled` and `proxy.disabled`.
+- Support `extraEnvs` for all control plan components
+- `extraArgs` changed from a map to structured extra arguments for duplicates
+- Add a `timeouts` structure for init, join, upgrade and reset.
+
+For details, you can see the official document below:
+
+- Support custom environment variables in control plane components under `ClusterConfiguration`.
+Use `apiServer.extraEnvs`, `controllerManager.extraEnvs`, `scheduler.extraEnvs`, `etcd.local.extraEnvs`.
+- The ResetConfiguration API type is now supported in v1beta4. Users are able to reset a node by passing
+a `--config` file to "kubeadm reset".
+- `dryRun` mode in is now configurable in InitConfiguration and JoinConfiguration.
+- Replace the existing string/string extra argument maps with structured extra arguments that support duplicates.
+The change applies to `ClusterConfiguration` - `apiServer.extraArgs`, `controllerManager.extraArgs`,
+`scheduler.extraArgs`, `etcd.local.extraArgs`. Also to `nodeRegistrationOptions.kubeletExtraArgs`.
+- Add `ClusterConfiguration.encryptionAlgorithm` that can be used to set the asymmetric encryption algorithm
+used for this cluster's keys and certificates. Can be one of "RSA-2048" (default), "RSA-3072", "RSA-4096" or "ECDSA-P256".
+- Add `ClusterConfiguration.dns.disabled` and `ClusterConfiguration.proxy.disabled` that can be used to disable
+the CoreDNS and kube-proxy addons during cluster initialization. Skipping the related addons phases,
+during cluster creation will set the same fields to `false`.
+- Add the `nodeRegistration.imagePullSerial` field in `InitConfiguration` and `JoinConfiguration`, which
+can be used to control if kubeadm pulls images serially or in parallel.
+- The UpgradeConfiguration kubeadm API is now supported in v1beta4 when passing --config to "kubeadm upgrade" subcommands.
+Usage of component configuration for kubelet and kube-proxy, InitConfiguration and ClusterConfiguration is deprecated
+and will be ignored when passing `--config` to upgrade subcommands.
```

Comment:
This sentence is a little bit misleading. It reads like all component configuration are deprecated AND are ignored for upgrade commands.

Reply:
Try to make it clear
```
For upgrade subcommands, the usage of component configuration for kubelet and kube-proxy, as well as InitConfiguration and
ClusterConfiguration, is deprecated and will be ignored when passing `--config`.
```

Reply:
Okay, this revision is much better.